packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts

```diff
@@ -1,470 +1,630 @@
+                output: computeBalancerSellQuote(pool, amount),
+                input: amount,
+                fillData: { poolAddress: pool.id },
+            })),
+        );
+    }
+
+    public async getBalancerBuyQuotesAsync(
+        makerToken: string,
+        takerToken: string,
+        makerFillAmounts: BigNumber[],
+    ): Promise<Array<Array<DexSample<BalancerFillData>>>> {
+        const pools = await this.balancerPoolsCache.getPoolsForPairAsync(takerToken, makerToken);
+        return pools.map(pool =>
+            makerFillAmounts.map(amount => ({
+                source: ERC20BridgeSource.Balancer,
+                output: computeBalancerBuyQuote(pool, amount),
+                input: amount,
+                fillData: { poolAddress: pool.id },
+            })),
+        );
+    }
+
+    public getMStableSellQuotes(
+        makerToken: string,
+        takerToken: string,
+        takerFillAmounts: BigNumber[],
+    ): SourceQuoteOperation {
+        return new SamplerContractOperation({
             source: ERC20BridgeSource.MStable,
-            encodeCall: contract => {
-                return contract
-                    .sampleSellsFromMStable(takerToken, makerToken, takerFillAmounts)
-                    .getABIEncodedTransactionData();
-            },
-            handleCallResultsAsync: async (contract, callResults) => {
-                return contract
-                    .getABIDecodedReturnData<BigNumber[]>('sampleSellsFromMStable', callResults)
-                    .map(amount => ({ amount }));
-            },
-        };
-    },
-    getMStableBuyQuotes(makerToken: string, takerToken: string, makerFillAmounts: BigNumber[]): SourceQuoteOperation {
-        return {
+            contract: this._samplerContract,
+            function: this._samplerContract.sampleSellsFromMStable,
+            params: [takerToken, makerToken, takerFillAmounts],
+        });
+    }
+
+    public getMStableBuyQuotes(
+        makerToken: string,
+        takerToken: string,
+        makerFillAmounts: BigNumber[],
+    ): SourceQuoteOperation {
+        return new SamplerContractOperation({
             source: ERC20BridgeSource.MStable,
-            encodeCall: contract => {
-                return contract
-                    .sampleBuysFromMStable(takerToken, makerToken, makerFillAmounts)
-                    .getABIEncodedTransactionData();
-            },
-            handleCallResultsAsync: async (contract, callResults) => {
-                return contract
-                    .getABIDecodedReturnData<BigNumber[]>('sampleBuysFromMStable', callResults)
-                    .map(amount => ({ amount }));
-            },
-        };
-    },
-    getMooniswapSellQuotes(
+            contract: this._samplerContract,
+            function: this._samplerContract.sampleBuysFromMStable,
+            params: [takerToken, makerToken, makerFillAmounts],
+        });
+    }
+
+    public async getBancorSellQuotesAsync(
```

Comment:
suggestion: rename/comment this similar to balancer to indicate it's pretty dangerous to call often.